Update cFE Filesystem

base0x10 opened the pull request:
### Summary of this Pull Request (PR)

This code provides all of the functionality required by the cFE.  It passes all of the unit tests, and further it implements all of the functionality which should be required by cFE applications.  It follows posix calls generally, but does not fully implement the behaviour of posix.  It also uses the osal API OS_ prefix for structs and functions rather than the posix names.  

The major changes of this PR include
- Implementing dynamic memory allocation for files
- Implementing functions for writing to or modifying files
- File descriptors for opening directories and files
- The tar parsing code has been moved into it's own file
- Input checking is done in a more sensible way
- Stat and Dirent structs are implemented temporarily by structs which follow the memory layout of glibc stat and dirent structs.  This is only until cFE is compiled with musl
- Permissions are implemented in a more sensible way

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [ x ] This PR is mature, and ready to be integrated into the repo. (Others/composite cFE2cos branch)

### Reviewers (Mandatory):
I would appreciate reviews @Others @zacharied  @gparmer 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

Style is a known issue in this pull request.  The style is consistent with all of the other code in the cFE2cos project.  This style is not consistent with either COS or cFE style and must be changed.  For this particular PR, most of the code will be adapted to adhere to COS style guidelines as a COS library.  The quality of the code in terms of readability, consistency, and maintainability is high.  

- [  ] Comments adhere to the Style Guide (SG)
- [  ] Spacing adhere's to the SG
- [  ] Naming adhere's to the SG
- [  ] All other aspects of the SG are adhered to, or exceptions are justified in this pull request

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [ x ] I've made an attempt to remove all redundant code
- [ x ] I've considered ways in which my changes might impact existing code, and cleaned it up
- [ x ] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [ x ] I've commented appropriately where code is tricky
- [ x ] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- cFE unit tests.  The outputs are summarized below.  
```
========================================================
ut_osfilesys TOTAL APIs: 11
ut_osfilesys TOTAL TEST CASES: 30

ut_osfilesys PASSED  25 tests.
ut_osfilesys FAILED   0 tests.

ut_osfilesys contains  1 tests that are untested OS-call-failure.
ut_osfilesys contains  0 tests that are manual-inspection-required.

ut_osfilesys contains  4 tests that are not-applicable.
========================================================
========================================================
ut_osfile TOTAL APIs: 24
ut_osfile TOTAL TEST CASES: 101

ut_osfile PASSED  78 tests.
ut_osfile FAILED   0 tests.

ut_osfile contains 20 tests that are untested OS-call-failure.
ut_osfile contains  0 tests that are manual-inspection-required.

ut_osfile contains  3 tests that are not-applicable.
========================================================
```

Note that OS-call-failure is not relevant to this platform.  They provide a mechanism to ensure that a failure within the OS call is reported to the program by the OSAL.  This code is not built on top of an OS filesystem which could fail.  

- cFE_booter was used to test this.  It is able to mount, open and read from the tar file system successfully
- debugging print statements were also used to test behaviour of functions in normal and edge cases.

base0x10 committed:
Implement a functional filesystem for the cFE

The cFE requires that the osal provide a nearly posix compliant
filesystem.  This includes abstractions for directory streams,
file descriptors, and physical volumes.  I have provided a
filesystem implementation which fulfills these abstractions and
implements all of the required API.

base0x10 committed:
fix filesystem directory api and style issues

Implemented many changes suggested during the pull request process.
Many of the changes are style or low impact changes.
The largest change is the directory API.  The directory stream
now behaves as expected, and a few methods have been moved so that
their implementations are all in one place.

base0x10 committed:
Merge branch 'cFE2cos' of https://github.com/Others/composite into cFE2cos

base0x10 committed:
Merge branch 'cFE2cos' of https://github.com/Others/composite into cFE2cos

base0x10 committed:
Update filesystem including good filedescripters, permissions, and file commands

The filesystem passes all of the cFE unit tests, it provides all of the
functionality that is required by the cFE_booter and that will be
needed by the linking and loading.  It is much closer to POSIX in the
way that is uses FDs.  It is still not threadsafe or modularized enough
to create a seperate library.

base0x10 committed:
Update name of tar function to read header

Old name of tar_cphdr was incorrect as there was no copying.
New name tar_hdr_read describes what the function does

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -9,6 +10,11 @@
+    uint32 ret = 0;
+    ret = newfs_init("/ramdev0", "RAM", 512, 4096);
+    if (ret != OS_FS_SUCCESS)   return ret;
+    ret = tar_parse();
```
Does tar_parse implicitly load the tar into the fs? This behavior, either way, should be documented in a comment.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    if (access != OS_READ_WRITE && access != OS_WRITE_ONLY) return OS_FS_ERROR;
+    if (ret != OS_FS_SUCCESS) return ret;
+    ret = file_create((char *) path, access);
+    return ret;
```
Don't bother with ret here:
`return file_create((char *) path, access);`

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    return read(filedes, buffer, nbytes);
+    if (chk_fd(filedes) != OS_FS_SUCCESS)  return OS_FS_ERR_INVALID_FD;
+    int32 ret = file_read(filedes, buffer, nbytes);
+    return ret;
```
No need for a ret variable here.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    int32 ret = path_chk_isvalid(path);
+    if (access != OS_READ_WRITE && access != OS_WRITE_ONLY) return OS_FS_ERROR;
+    if (ret != OS_FS_SUCCESS) return ret;
+    ret = file_create((char *) path, access);
```
You wrote `file_create`. You shouldn't have to cast the `const` away from path.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    }
+    int32 ret = path_chk_exists(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return file_open((char *)path, access);
```
You wrote `file_open`. You shouldn't have to cast the `const` away from path.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    int32 ret = path_chk_exists(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    ret = file_remove((char *)path);
+    return ret;
```
Just use `return file_remove((char *)path);` And you wrote `file_remove`, you should not have to cast away const!

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    if (ret != OS_FS_SUCCESS) return ret;
+    ret = path_chk_isvalid(new_filename);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return file_rename((char *)old_filename, (char *)new_filename);
```
What happens in the case where the new_filename is already used?

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    return 0;
+    if (!src || !dest) return OS_FS_ERR_INVALID_POINTER;
+    char *src_path  = (char *)src;
+    char *dest_path = (char *)dest;
```
You wrote all these functions you're using! Stop casting away const!

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    ret = path_chk_isvalid(dest_path);
+    if (ret != OS_FS_SUCCESS) return ret;
+
+    return file_cp(src_path, dest_path);
```
What's the difference between rename and copy?

base0x10 replied:
Copy moves a file to a new location, possibly within a new directory.  Also leaves one of the copies in the original path.  exact same behaviour as linux/bash/gnu coreutils

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Read Only Filesystem!");
-    return 0;
+    char *src_path  = (char *)src;
+    char *dest_path = (char *)dest;
```
Casting away const is an anti-pattern

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    uint32 i;
+    for (i = 1 ; i < MAX_NUM_FILES + 1 ; i++) {
```
`i <= MAX_NUM_FILES`

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    return 0;
+    int32 ret = path_chk_isvalid(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return file_mkdir((char *)path);
```
Don't cast away const when passing an argument to functions that you wrote

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if(path_chk_exists(path) != OS_FS_SUCCESS) return NULL;
+    int32 FD = dir_open((char *)path);
```
You're casting away const here for no reason.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (!directory) return OS_FS_ERR_INVALID_POINTER;
+    return dir_close((int32) directory);
```
Shouldn't `dir_close` just take a `dirp_t`?

base0x10 replied:
dirp is a pointer of some kind that refers to the dir's underlying structure.  In our case (and this is fairly standard) the underlying structure is a fd, and that dir pointer is the offset into the fd table.  Unless you require the posix wrapper, people who want to use dir_close will use the fd which is just an integer.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
+    if (!directory) return;
+    dir_rewind((int32) directory);
```
Can dir_rewind just take a `dirp_t`?

base0x10 replied:
The level of abstraction for the dir_functions is file descriptors which are integer offsets into a table.  This is all that the simplest level should use.  posix/cfe require these wrapper structs.  see comment about closedir

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (!directory) return NULL;
```
Shouldn't `dir_read` handle this for you?

base0x10 replied:
It does and I am removing this line for now.  But I would like to talk further on this.  the API spec says return null if directory is null.  It seems more maintainable if I check for what the api considers invalid inputs at this level, and then check for what the underlying file system considers a valid input at the filesystem level.  That way changes can be made to the filesystem without breaking the API level.  Thoughts?

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (!directory) return NULL;
+    os_dirent_t *dir = dir_read((int32) directory);
```
Can `dir_read` just take a `os_dirp_t`? This seems obvious possible, since it returns a os_dirent_t* anyway.

base0x10 replied:
At the lowest level of the fs library there is no reason to need to use the posix pointer dirp_t.  It doesnt provide anything besides wrapping a fd.  however dirent provides a unique functionality.  perhaps ultimately I will use a simplified dirent structure for the lowest level of the fs library and then make the posix call return a full dirent.  for now dirent is needed so the lower level api must deal with it, while dirp_t is not needed so the lower level does not deal with it

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    return 0;
+    int32 ret = path_chk_exists(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return file_rmdir((char *)path);
```
Casting away const is code smell

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -260,8 +320,8 @@ int32 OS_initfs(char *address, char *devname, char *volname, uint32 blocksize, u
 */
```
Can't comment on it directly, since it's not modified, but can OS_initfs just always return OS_ERR_UNIMPLEMENTED?

base0x10 replied:
As far as I can tell this is used to set up some space for an existing or new filesystem (on disk?).  It looks like it can also be used to initialize a preloaded filesystem.  We don't need to do anything special to prepare for an existing fs to be mounted, or a new fs to be created.  So if it is called there is no reason to return anything other than success in case an application relies on it returning successful.

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -270,16 +330,23 @@ int32 OS_rmfs(char *devname)
-    return unmount(mountpoint);
+    int32 ret = path_chk_exists(mountpoint);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return fs_unmount((char *)mountpoint);
```
Casting away const literally makes some children cry

zacharied replied:
@Others Boosting your comment count i see

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -299,17 +366,18 @@ int32 OS_fsBytesFree(const char *name, uint64 *bytes_free)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    return OS_ERR_NOT_IMPLEMENTED;
```
Can we actually do this, since it expects a `os_fshealth_t ` to be returned?

zacharied replied:
FWIW `OS_ERR_NOT_IMPLEMENTED = -28`.

base0x10 replied:
lol, negative health.  This is the declaration for the type returned.  
```
/* still don't know what this should be*/
241 typedef unsigned long int   os_fshealth_t;
```

base0x10 replied:
So you think I should just return unsigned long max value?

base0x10 replied:
linux returns unimplemented

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -318,17 +386,16 @@ int32 OS_FS_GetPhysDriveName(char * PhysDriveName, char * MountPoint)
 {
-    strcpy(LocalPath, VirtualPath);
-    return OS_FS_SUCCESS;
+    return path_translate((char *) VirtualPath, LocalPath);
```
Casting away const is like casting away human dignity

Others commented on src/components/implementation/no_interface/cFE_booter/tar.h:
```diff
@@ -0,0 +1,34 @@
+
+uint32 round_to_blocksize(uint32 offset);
+
+uint32 oct_to_dec(char *oct);
```
Do these need to be exposed?

zacharied replied:
Maybe leave this out and declaring its definition as `static`.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+//should be overwritten by linking step in build process
+__attribute__((weak)) char _binary_cFEfs_tar_size=0;
+__attribute__((weak)) char _binary_cFEfs_tar_start=0;
+__attribute__((weak)) char _binary_cFEfs_tar_end=0;
```
Is the `=0` necessary here? Isn't it implicitly 0?

zacharied replied:
I think it helps a little bit to declare the default value of a `weak`.

Others replied:
@zacharied but it's pointless here. In C, globals always default to 0.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+    for (base = 1, tot = 0 ; i >= 0 ; i--, base *= 8) {
+        char val = oct[i];
+        assert(val <= '7' && val >= '0');
+        val = val - '0';
```
This might be a little too clever... But w/e it obviously works if you think about it.

base0x10 replied:
This is gparmer's code

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+ * the linker are non-intuitive so a description of error checking is given:
+ * First checks to make sure that symbols have been overwritten by linking
+ * process.  Next checks that the size is greater than 0.  Finally checks that
+ * the end of the tar is after the start
```
This is a great comment, but I would inline the comments with the checks they refer to. Otherwise it's too easy for the to get out of sync.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+//locations and size of tar
+char      *tar_start;
+char      *tar_end;
+size_t    tar_size;
```
This all relies pretty heavily on global data, could future work be to pass in parameters? Probably just needs a TODO comment for now though.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+    uint32 offset = 0;
+    struct fsobj *o;
+
+
```
One blank line is fine here I think.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+            return OS_FS_SUCCESS;
+        }
+        if (tar_hdr_read(offset, o))           return OS_FS_ERR_DRIVE_NOT_CREATED;
+        if (file_insert(o, offset + tar_start)) return OS_FS_ERR_DRIVE_NOT_CREATED;
```
Check explicitly against success?

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+
+/*
+ * Copies information from a tar file header to a fsobj
+ * TODO: change name of this function and make it cast to tar hdr struct
```
Fix these comments, they're out of sync with the code.

Others commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+    char *location = tar_start;
+    location += tar_offset;
+    memmove(location + 1, location, strlen(location));
+    location[0] = '/';
```
This code is quite difficult to parse, some spacing and or comments would go a long way.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
 
+#include <cos_component.h>
+#include <cos_defkernel_api.h>
+#include <cos_kernel_api.h>
```
This includes should go above the `gen/*` includes

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
-#define TAR_BLOCKSIZE 512
-#define INT32_MAX 0x7FFFFFF //2^31 - 1
+// a page is 4096, size of f_part is 5 values * 4 bytes
+#define F_PART_DATA_SIZE (4096-sizeof(struct f_part))
```
Can we use the PAGE_SZ constant here?

base0x10 replied:
did you mean PAGE_SIZE? PAGE_SZ isnt used.  also is it worth including shared consts for just that and does the bump pointer always return 4096 or does it return PAGE_SIZE

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
     FSOBJ_FILE,
     FSOBJ_DIR,
-} fsobj_type_t;
+} fsobj_type; //TODO: what is the preferred naming convention for this
```
Don't use a typedef here, just use `enum fsobj_type` everywhere

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+    WRITE,
+    READ_WRITE,
+    NONE,
+} permission_t;
```
Don't use a typedef here, use `enum fs_ permissions` here. Also, maybe this could follow the `|` convention? Ex:
```
enum fs_permissions {
  NONE =    0,
  READ =    0x001,
  WRITE =   0x010,
  EXECUTE = 0x100,
  ALL = READ | WRITE | EXECUTE,
};
```

zacharied replied:
Better yet just use `1 << x` where `x` is the enum index

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+typedef enum {
+    STATIC,
+    DYNAMIC,
+} fpart_alloc_t;
```
Don't use `typedef` here. Try `enum fpart_alloc_type`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+    struct fsobj *file;
+    struct f_part *next, *prev;
+    char *data;
 };
```
Such a good use-case for `ps_list`

base0x10 replied:
true, but that is not something I want to change in this PR.  I think it is still worth putting into the repo before we switch to PS list

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+    CUR_DIR_LINK,
+    PARENT_DIR_LINK,
+    END_OF_STREAM,
+} dir_stream_status_t;
```
Don't use `typedef` here, prefer `enum dir_stream_status`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+int32 file_mv(char *src, char *dest);
 
-uint32 file_insert(struct fsobj *o, struct FS *fs);
+int32 chk_fd(int32 FD);
```
What does this even do? Consider renaming.

base0x10 replied:
I updated with a comment where it is implemented but I don't think it is that ambiguous or worth changing for this pr

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+int32 chk_fd(int32 FD);
 
-uint32 insert(struct fsobj *o, struct fsobj *root);
+int32 file_FDGetInfo(int32 FD, OS_FDTableEntry *fd_prop);
```
Inconsistent naming. I know this is probably cFE convention, but this is composite specific stuff. Try `file_fd_info`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+int32 file_lseek(int32 FD, int32 offset, uint32 whence);
 
-int32 read(int32  filedes, void *buffer, uint32 nbytes);
+permission_t permission_cFE_to_COS(uint32 permission);
```
This name is just confusing. How about `file_permission_from_cFE_permission`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+permission_t permission_cFE_to_COS(uint32 permission);
 
-struct fsobj *find(const char *path, struct fsobj *root);
+uint32 permission_COS_to_cFE(permission_t permission);
```
How about `file_permission_to_cFE_permission`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+
+os_dirent_t *dir_read(int32 FD);
+
+int32 file_mkdir(char *path);
```
I don't think I like this name. Would `dir_create` be so bad?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+
+int32 file_mkdir(char *path);
+
+int32 file_rmdir(char *path);
```
I don't think I like this name. Would `dir_remove` be so bad?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
 /******************************************************************************
 ** fs Level Methods
 ******************************************************************************/
+int32 path_chk_isvalid(const char *path);
```
`path_check_is_valid`, or maybe even `path_verify`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
 ******************************************************************************/
+int32 path_chk_isvalid(const char *path);
+
+int32 path_chk_exists(const char *path);
```
Honestly `path_exists` is probably fine.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+int32 file_rmdir(char *path);
 
 /******************************************************************************
 ** fs Level Methods
```
I'd split this into two sections. Path related methods, and fs related methods.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
-uint32 fs_init(struct FS *filesys, char *devname, char *volname, uint32 blocksize, uint32 numblocks);
+uint32 fs_init(struct fs *filesys, char *devname, char *volname, uint32 blocksize, uint32 numblocks);
 
 uint32 newfs_init(char *defilvname, char *volname, uint32 blocksize, uint32 numblocks);
```
Having two functions, one named `fs_init` and one called `newfs_init` is just confusing.

zacharied replied:
If this is totally necessary, put some comments to explain the difference.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
 
 uint32 newfs_init(char *defilvname, char *volname, uint32 blocksize, uint32 numblocks);
+
+int32 rmfs(char *devname);
```
I think `fs_remove` is probably better.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+
+int32 rmfs(char *devname);
+
+int32 filesys_GetPhysDriveName(char *PhysDriveName, char *MountPoint);
```
I know this is cFE convention, but I'd go with `fs_get_physical_drive_name`. (And the arguments definitely should be snake_case)

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+
+int32 filesys_GetPhysDriveName(char *PhysDriveName, char *MountPoint);
+
+int32 filesys_GetFsInfo(os_fsinfo_t *finesys_info);
```
See above. Go with `fs_get_info`. (At the very least stick with just `fs`, not `fs` and `filesys`.) Also the argument name is misspelled.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
 
-struct FS fs[MAX_NUM_FS];
+struct fs filesystems[MAX_NUM_FS];
+struct fs *openfs = &filesystems[0];;
```
Extra `;`

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+struct cos_compinfo *ci;
+
+struct fd fd_tbl[OS_MAX_NUM_OPEN_FILES + 1];
+char copy_buffer[F_PART_DATA_SIZE];
```
Honestly I'd put this near where it is used. (This may be the very rare case where making a global into a static variable in a function might improve clarity.)

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (!filedes->file) return OS_FS_ERR_INVALID_FD;
+    if (filedes->ino != filedes->file->ino) return OS_FS_ERR_INVALID_FD;
```
This feels like a critical failure, so maybe it should be an assert? Or is there some case where this can happen?

zacharied replied:
Probably better to let the cFE or app handle it, no? More flexibility.

Others replied:
@zacharied but, in this case, doesn't it always mean an invariant has been broken? AKA there is a bug in the FS implementation?

base0x10 replied:
it would be a legal state of the filesystem if one of both of them were 0 and the other a valid other value.  The real invalid state would be if file->ino and filedes->ino were nonezero and different.  I now explicitly test for both of these and do the right thing in each case.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
- */
-uint32 tar_parse()
+//finds the next free file
+uint32 newfile_get(struct fsobj **o)
```
I'm not happy with this name, perhaps `file_get_new`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    if (!fs[0].devname) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    **o = (struct fsobj) {
+        //ino needs to be unique and nonzero, so ino is defined as index+1
+        .ino = count +1
```
Operator spacing!

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+uint32 file_insert(struct fsobj *o, char *path)
+{
+    //paths should always begin with '/' but we do not need it here
+    if (path[0] != '/') return OS_FS_ERR_PATH_INVALID;
```
The check if path is null happens **after** you use it. -_-

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    while (offset + start < end) {
-        //tar ends after two empty records
-        if ( !(offset + start)[0] && !(offset + start)[TAR_BLOCKSIZE]) return OS_FS_SUCCESS;
+    assert(o && path && openfs);
```
I think when `o` or `path` is null we should probably return an error code instead.

base0x10 replied:
I am in a weird place.  That checking is done in the wrapper methods (as it should be), and redundant code is not good.  When this is a library it will return error codes, and the cFE will interpret them to return the cFE error codes.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    char *token = strtok(path_temp, delim);
+
+    if (strcmp(token, cur->name) != 0) return OS_FS_ERR_PATH_INVALID;
+    //loop terminates when it finds a place to put o or determines path is invalid
```
Move comment down a line.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    // token is the current directory in path
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
```
No need for number in brackets here I think.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
```
Isn't this non-reentrant?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            }
+            cur = cur->next;
+        }
+    }
```
This loop is kinda crazy. Can this be broken up?

zacharied replied:
It's commented well though

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+struct fs filesystems[MAX_NUM_FS];
+struct fs *openfs = &filesystems[0];;
 struct fsobj files[MAX_NUM_FILES];
+struct cos_compinfo *ci;
```
There is a lot of global data here. It might be better, when this becomes a library, to make the user pass in this data. (Perhaps as a pointer.) As an intermediate change, it might be best to turn this into a single global struct? Idk, ymmv.

zacharied replied:
Many C libraries have a something like a `connection` datatype that represents an interface to the library and is passed to any calls to that library.

base0x10 replied:
A global struct sounds like a great idea to pass the context to the fs, but I think it adds complexity without improving much at this stage.  Will be done when broken up into libraries

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+int32 file_open(char *path, permission_t permission)
+{
+    assert(openfs);
+    if (!openfs->root) return OS_FS_ERR_PATH_TOO_LONG;
```
This error code seems bizarre. Why would laking a root imply the path is too long???

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
 ******************************************************************************/
 
-uint32 fsobj_init(struct fsobj *o)
+uint32 newpart_get(struct f_part **part)
```
`part_get_new`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+******************************************************************************/
+
+// Internally, FDs are considered unused when ino == 0
+int32 newfd_get(int32 ino)
```
`fd_get_new`?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    assert(tar_offset < size);
-    assert(file->filedes > 0);
+    int32 ret = chk_fd(FD);
+    if (ret != OS_FS_SUCCESS) return OS_FS_ERROR;
```
Can we not just return `ret`?

base0x10 replied:
chk_fd returns invalid fd code, but the api specifies that close will only return success or error.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            return WRITE;
+        case OS_READ_ONLY :
+            return READ;
+        default :
```
Can we use an explicit case for this? And then make the default assert? This should improve future compatibility.

base0x10 replied:
We can make the default assert, but NONE cannot be an explicit case as cFE does not provide any access code that maps to none

Others replied:
:+1: default should assert

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            return OS_WRITE_ONLY;
+        case READ :
+            return OS_READ_ONLY;
+        default :
```
What about permission NONE?

base0x10 replied:
the cFE does not provide any permission which maps to NONE.  Data corruption or something else screwy is going on if a file it has access to is not read, write, or read & write

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+{
+    int32 ret = path_chk_isvalid(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    if (file_find((char *)path) == NULL) return OS_FS_ERROR;
```
Don't cast away const

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    int32 ret = path_chk_isvalid(path);
+    if (ret != OS_FS_SUCCESS) return ret;
+    if (file_find((char *)path) == NULL) return OS_FS_ERROR;
+    return 0;
```
Return success explicitly

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    if (!ci) {
+        struct cos_defcompinfo *defci = cos_defcompinfo_curr_get();
+        ci = &defci->ci;
+    }
```
Do you really need to cache ci globally? Why can't we just fetch it every time?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    (*part)->next = NULL;
+    (*part)->prev = NULL;
+    (*part)->file = NULL;
+    (*part)->data = (char *) *part + sizeof(struct f_part);
```
The way you are using `fpart` here is literally the DST pattern. Therefore I think you should make `fpart` a dynamically sized type, with a flexible array member at the end.

base0x10 replied:
flexable array member cannot work in this case, as the data in a static file (TAR loaded file) is not sequentially right after the f_part struct for that file

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    struct file_position *position = &filedes->position.file_pos;
+    struct f_part *part = position->open_part;
+    if (o->refcnt < 1)                              return OS_FS_ERROR;
+    if (o->type == FSOBJ_DIR)                       return OS_FS_ERROR;
```
You check against o->type twice. This is redundant.

zacharied replied:
Condense these two `if`s into one

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    //filedes needs to be unique and nonzero, so filedes is defined as index+1
-    return OS_FS_SUCCESS;
+
+    if (nbytes == 0)            return 0;
```
Return success explicitly

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    return OS_FS_SUCCESS;
+
+    if (nbytes == 0)            return 0;
+    uint32 bytes_to_read        = nbytes;
```
What is this aligned with?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    uint32 bytes_to_read        = nbytes;
+
+    if (o->memtype == DYNAMIC){
+        uint32 read_size; //the length of a continuous segment to be read from
```
Put this in the scope where it is used.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+                position->part_offset     += bytes_to_read;
+                position->file_offset     += bytes_to_read;
+                return nbytes;
+            }
```
Simplify! I think this is too complicated.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+        return nbytes;
+    } else {
+        PANIC("Memtype is neither static nor dynamic");
+    }
```
Invert these ifs, so you don't have as deep nesting.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
 {
-    assert(filesys && filesys->root && filesys->root->filedes);
-    if (filesys->root == o) return OS_FS_SUCCESS;
+    assert(path);
```
Return an error code for consistency.

base0x10 replied:
It seems logical to return a null to a function that returns a file if that file cannot be found.  Idk, does that make sense, or is there something i'm missing

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
```
This is racy

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
+
```
The above code looks like a duplicate of code elsewhere where you use strtok. DRY violation.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+        }
+        token = strtok(NULL, delim);
+        if (token == NULL) return cur;
+        if (!cur->child) return NULL;
```
Lift this check into the while condition, and then return NULL at the end of the function. Also consider a for loop to do this iteration.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    o->name = path_to_name(path);
+    o->type = FSOBJ_FILE;
+    o->size = 0;
+    o->permission = permission;
```
Can you use a designated initializer here?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+
+    struct f_part *part;
+    newpart_get(&part);
+    assert(part);
```
Can't you just check the return code?

zacharied replied:
I think we should stray away form `assert`s in general and use the built in cFE error codes.

Others replied:
@zacharied I think if the violation of a constraint always means there is a bug, then asserts are fine. Otherwise, I agree if you.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    o->memtype = DYNAMIC;
+    o->file_part = part;
+    part->file = o;
+    part->data = (char *) part + sizeof(struct f_part);
```
Again, fpart could be a DST

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+{
+    struct fsobj *file = file_find(path);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    file_rm(file);
```
Can file_rm not fail?

base0x10 replied:
that is correct

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    assert(strlen(root->name) < strlen(o->name));
+    struct fsobj *file = file_find(old_filename);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    file->name = path_to_name(new_filename);
```
What happens it the there already is a file at `new_filename`?

base0x10 replied:
the existance of a file of the same name as new_filename is checked for in the wrapper layer

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+        filestats->st_mode = S_IFDIR;
+    }
+    filestats->st_size = file->size;
+    filestats->st_blksize = F_PART_DATA_SIZE;
```
Most of this filestats stuff can be a designated initializer. Also, please document where this differs from a posix implementation.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    if (whence == SEEK_SET) {
+        if (offset < 0) return OS_FS_ERROR;
+        target_offset = offset;
+
```
Extra space

zacharied replied:
Get that outta here

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    return OS_FS_SUCCESS;
+}
+
+int32 file_lseek(int32 FD, int32 offset, uint32 whence)
```
Should whence be an enum?

base0x10 replied:
I don't define whence, posix does

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    if (position->file_offset > o->size) {
+        o->size = position->file_offset;
+    }
+    assert(position->file_offset == target_offset);
```
Why do we need to variables here, if they're supposed to be in lockstep?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    int32 to_copy = fd_tbl[fd_src].file->size;
+    int32 read_size = 0, write_size = 0;
+
+    // TODO: copy buffer is aggressively not thread safe, take a lock
```
Most of the code in this file isn't thread safe

base0x10 replied:
yeah.... I know....

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    file_close(fd_dest);
+
+    return OS_FS_SUCCESS;
+}
```
Is it an issue that copy uses 2 of the available file descriptors?

base0x10 replied:
no...

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+ */
+int32 file_FDGetInfo(int32 FD, OS_FDTableEntry *fd_prop)
+{
+
```
Extra space

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    struct dir_position *position = &filedes->position.dir_pos;
+    position->open_dir = file;
+    position->cur = file;
+    position->status = NORMAL;
```
Designated initializer?

base0x10 replied:
I think the way it is, is clear.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+
+    struct fd *filedes = &fd_tbl[FD];
+    filedes->access = READ;
+    filedes->file = file;
```
Designated initializer?

base0x10 replied:
seems excessive, im only directly modifying two fields.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    return nbytes;
+void dir_rewind(int32 FD)
+{
+    if (FD > OS_MAX_NUM_OPEN_FILES + 1 || FD <= 0) return;
```
Isn't this just ignoring the failing case? Shouldn't we do something here? Return an error code?

base0x10 replied:
Rewind does not have anyway of failing.  its only guaranteed to rewind a file if it is passed a valid FD

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    struct fd *filedes = &fd_tbl[FD];
+
+    if (filedes->ino == 0) return;
+    if (filedes->file->type != FSOBJ_DIR) return;
```
See above

base0x10 replied:
see above

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    o->prev = NULL;
+    o->parent = NULL;
+    o->child = NULL;
+    o->size = 0;
```
Designated initializer?

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            // This is a bad hack that I have not found a solution to
+            // basically mount should fail if already mounted
+            // but to load tar I need to pre-mount the first filesystem
+            if (strcmp(mountpoint, "/ram") && filesystems[i].root) return OS_FS_ERROR;
```
This sucks, but I can't think of a better solution

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    filesys->root           = o;
-    filesys->next           = NULL;
-    filesys->prev           = NULL;
+    filesys->root           = NULL;
```
Designated initializer? That way you can leave off the NULL ones.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
+            filesystems[i].mountpoint = NULL;
+            filesystems[i].blocksize = 0;
+            filesystems[i].numblocks = 0;
+            filesystems[i].root = NULL;
```
Designated initializer!

base0x10 replied:
I am relying on not zeroing out fields that I dont modify.

Others commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
+    uint32 i;
+    for (i = 0 ; i < MAX_NUM_FS && filesystems[i].devname != NULL ; i++) {
+        if (filesystems[i].mountpoint && !strcmp(filesystems[i].mountpoint,MountPoint)) {
+            memcpy(PhysDriveName, "RAM FS\n",7);
```
Kinda sketchy that the size is hardcoded like this. Is there a size of trick you could use here?

Others replied:
strcopy

Others reviewed (changes requested):
This is great! I'm very happy that it's finally ready to get merged. (Although it could've been split up into smaller PRs initially...) There are quite a few notes I have on this, but they all should simple fixes. Great work!

Others commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
+    ret = path_chk_isvalid(dest_path);
+    if (ret != OS_FS_SUCCESS) return ret;
+
+    return file_cp(src_path, dest_path);
```
Can we get a comment for this?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -9,6 +10,11 @@
+    ret = newfs_init("/ramdev0", "RAM", 512, 4096);
+    if (ret != OS_FS_SUCCESS)   return ret;
+    ret = tar_parse();
+    if (ret != OS_FS_SUCCESS)   return ret;
```
Can't this just be changed to `return ret;`?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
-    PANIC("Unimplemented method!"); // TODO: Implement me!
-    return 0;
+    if (!fd_prop)                                   return OS_FS_ERR_INVALID_POINTER;
+    if (filedes <= 0 || filedes > MAX_NUM_FILES)    return OS_FS_ERR_INVALID_FD;
```
Perhaps leave a `print` here that differentiates between whether `filedes` is negative versus too high — it might make things easier to debug.

Others replied:
@zacharied we can't leave debug prints in this kind of code. Logging would be great though...

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -17,201 +23,260 @@ int32 OS_FS_Init(void)
 void OS_rewinddir(os_dirp_t directory)
 {
-    PANIC("Unimplemented method!"); // TODO: Implement me!
+    if (!directory) return;
```
Consider handling this in `dir_rewind`.

base0x10 replied:
it is also handled in dir_rewind when dir_rewind checks that the file descriptor is within the range of valid file descriptors.  But I will remove it here.

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -220,27 +285,22 @@ int32 OS_rmdir(const char *path)
  * Makes a file system
 */
-int32 OS_mkfs(char *address,char *devname, char *volname,
+int32 OS_mkfs(char *address, char *devname, char *volname,
```
Nice catch Joe!

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -220,27 +285,22 @@ int32 OS_rmdir(const char *path)
-    if (ret != OS_FS_SUCCESS)   return ret;
-    return OS_FS_SUCCESS;
+                                return OS_FS_ERR_PATH_TOO_LONG;
+    return newfs_init(devname, volname, blocksize, numblocks);
```
What is `newfs`? Sorry if you already explained it.

base0x10 replied:
It inits a new fs.  by popular demand I think I will change around the naming to this and others

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -260,8 +320,8 @@ int32 OS_initfs(char *address, char *devname, char *volname, uint32 blocksize, u
-    PANIC("Read Only Filesystem!"); // TODO: Implement me!
-    return 0;
+    if (!devname) return OS_FS_ERR_INVALID_POINTER;
+    return rmfs(devname);
```
`fs_rm`?

base0x10 replied:
good catch, thanks

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -299,17 +366,18 @@ int32 OS_fsBytesFree(const char *name, uint64 *bytes_free)
+    if (!PhysDriveName || !MountPoint) return OS_FS_ERR_INVALID_POINTER;
+    int32 ret = path_chk_exists(MountPoint);
+    if (ret != OS_FS_SUCCESS) return ret;
+    return filesys_GetPhysDriveName(PhysDriveName, MountPoint);
```
`fs`?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -220,27 +285,22 @@ int32 OS_rmdir(const char *path)
 {
     if (!devname || !mountpoint) return OS_FS_ERR_INVALID_POINTER;
-    return mount(devname, mountpoint);
+    return fs_mount((char *)devname, mountpoint);
```
@Others missed this one

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfiles.c:
```diff
@@ -270,16 +330,23 @@ int32 OS_rmfs(char *devname)
+    * if we expect the filesystem to be used much, that is the right choice, but
+    * otherwise it is not worth actually doing
+    */
+    return OS_ERR_NOT_IMPLEMENTED;
```
I assume this method isn't called very often if we're not implementing it?

base0x10 replied:
I don't know.  I don't think it will be called but I need to find applications and search through them.  I grep'd our code base.  none of the applications we have downloaded call it.  But it is called once by ` cfe/fsw/cfe-core/src/es/cfe_es_start.c ` which means I should probably implement it.

base0x10 replied:
So do you guys think imposing an artificial limit on the amount of space being used is the right choice.  And then just count the number of bump page allocations? @Others  as well

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+        //if there is no child, then insert as child
+        if (!cur->child) {
+            // if the next part of the path is not o->name or there is a part after it, bad path
+            token = strtok(NULL, delim);
```
watch out my dude

base0x10 replied:
I don't understand?  But it doesn't matter because I am getting rid of strtok anyway.  But this is a correct use of strtok

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            PANIC("Invalid permission in existing file");
+    }
+    PANIC("unreachable statement");
+    return 0;
```
What's up with this

base0x10 replied:
This is the pattern that @Others recommended at last pr for unreachable but necessary return statements.  But im down if there is a better pattern

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+
+                buffer                      += read_size;
+                bytes_to_read               -= read_size;
+                position->file_offset       += read_size;
```
This much alignment is hard to read IMO - why isn't it closer?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    if (o->memtype == STATIC)           return OS_FS_ERROR;
+    if (o->type == FSOBJ_DIR)           return OS_FS_ERROR;
+
+    if (nbytes == 0)                    return 0;
```
More alignment shenanigans

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+    }
+
+    file->name = path_to_name(dest);
+    file_insert(file, dest);
```
`file_insert` has no return code?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    return nbytes;
+void dir_rewind(int32 FD)
+{
+    if (FD > OS_MAX_NUM_OPEN_FILES + 1 || FD <= 0) return;
```
`FD >= OS_MAX_NUM_OPEN_FILES`

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+int32 file_rm(struct fsobj *o)
+{
+    // I think I should keep a file from being deleted if it is open, rather than closing it
+    //file_close_by_ino(o->ino);
```
At least put a log message here if it's non-standard behavior

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
+            struct fsobj *o;
+            if (newfile_get(&o))        return OS_FS_ERROR;
+            filesystems[i].mountpoint   = mountpoint;
+            openfs                      = &filesystems[i];
```
Be sure to keep consistent for alignment in assignments.

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -1,280 +1,886 @@
-    for (i=0 ; i < MAX_NUM_FS ; i++) {
-        if (mountpoint != NULL && !strcmp(fs[i].mountpoint, mountpoint)) {
-            fs[i].mountpoint = NULL;
+    assert(mountpoint);
```
Just return an error here instead?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
         ret = tar_load();
         if (ret != OS_FS_SUCCESS) return ret;
-        ret = fs_init(&fs[0], devname, volname, blocksize, numblocks);
+        ret = fs_init(&filesystems[0], devname, volname, blocksize, numblocks);
```
What is the difference between `newfs_init` and `fs_init`? Some name changing might be in order

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
+
+}
 
+int32 filesys_GetPhysDriveName(char *PhysDriveName, char *MountPoint)
```
There are only two hard things in Computer Science: cache invalidation and naming things.

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.c:
```diff
@@ -286,23 +892,73 @@ uint32 newfs_init(char *devname, char *volname, uint32 blocksize, uint32 numbloc
+    return OS_FS_ERROR;
+}
+
+int32 filesys_GetFsInfo(os_fsinfo_t *filesys_info)
```
Watch name prefix and capitalization

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
-size_t    size;
+// for exactly the same reason as above, we need to declare a stat structure to pass
+// back to the cFE that it will recoginize
+struct hack_stat
```
Consider putting TODO's everywhere `hack_*` symbols are used — should be a pretty simple `grep`.

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+int32 file_close_by_ino(int32 ino);
 
-uint32 tar_parse();
+int32 file_close_by_name(char *path);
```
`const char*` perhaps?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+uint32 newfile_get(struct fsobj **o);
 
-uint32 round_to_blocksize(uint32 offset);
+uint32 file_insert(struct fsobj *o, char *path);
```
`const char*` perhaps?

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
 
-    return tot;
-}
+struct fsobj *file_find(char *path);
```
`const`

zacharied commented on src/components/implementation/no_interface/cFE_booter/osfilesys.h:
```diff
@@ -1,113 +1,258 @@
+/******************************************************************************
+** dirent Level Methods
+******************************************************************************/
+int32 newfd_get(int32 ino);
```
Getting rather confusing, have you considered just prefixing everything with `fs_`?

zacharied commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+}
+
+//used to convert filesize in oct to dec
+uint32 oct_to_dec(char *oct)
```
Maybe put a `TODO` for just grabbing a function like this from an external math library

base0x10 replied:
This wouldnt be in a math lib, it converts from an ascii string representing octal values to an int.

zacharied commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+    int32 i, base;
+    int32 tot;
+    i = strlen(oct) - 1;
+    for (base = 1, tot = 0 ; i >= 0 ; i--, base *= 8) {
```
Give a comment

zacharied commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+ */
+uint32 tar_hdr_read(uint32 tar_offset, struct fsobj *file)
+{
+    assert(tar_offset < tar_size);
```
Maybe use cFE error handling

zacharied reviewed (changes requested):
Looks pretty functional — most comments are just in the interest of maintainability.

base0x10 commented on src/components/implementation/no_interface/cFE_booter/tar.c:
```diff
@@ -0,0 +1,128 @@
+}
+
+//used to convert filesize in oct to dec
+uint32 oct_to_dec(char *oct)
```
Maybe I need a comment.  it translates from string of ascii characters that represent 1-8 that represent an octal value, and converts to int type.  I dont think math libraries would have this

base0x10 committed:
Merge branch 'cFE2cos' of https://github.com/Others/composite into cFE2cos

base0x10 committed:
Update to file system for musl libc and minor changes

We are now compiling cFE under musl libc.  This broke the temporary hack
we were using to allow us to pass structs between musl libc and cFE's
glibc
Also I have made style changes, including comments and function names.
Permissions are now implemented as bits in the permission enum, so they
can use & and | notation.

base0x10 committed:
Tiny style changes in cos osal cFE code.

change COS to cos for style reasons, and panic if an invalid
value is passed as a cFE permission_cFE_to_cos
